using basic authentication, cannot login into the application using taiko and gauge

### Expected behavior
Using in built basic authentication mechanism of taiko, the user id and password should be filled in and able to login into the website.

### Actual behavior
It fails stating that there is an authentication error. I am able to login when I manually launch the url and provide the credentials
> gauge-taiko-template@0.0.1 test.u d:\ProgramData\BondDevSB\TrunkServices\Bidscout_Pipeline\CI_CD\e2e_tests\cs-bidscout-web-config
> gauge run specs/underwriter.spec

**Underwriter can use BidScout for Underwriters**
(node:11660) ExperimentalWarning: The fs.promises API is experimental
[DEPRECATED] gauge.screenshotFn will be removed soon, use gauge.customScreenshotWriter instead.

 ****Error Message: Error: Navigation to url http://uw-bidscout-dev-basic-auth.dvllb.travp.net/bidscout-underwriter failed. REASON: net::ERR_INVALID_AUTH_CREDENTIALS
  Stacktrace: 
  Error: Navigation to url http://uw-bidscout-dev-basic-auth.dvllb.travp.net/bidscout-underwriter failed. REASON: net::ERR_INVALID_AUTH_CREDENTIALS****

Successfully generated html-report to => d:\ProgramData\BondDevSB\TrunkServices\Bidscout_Pipeline\CI_CD\e2e_tests\cs-bidscout-web-config\reports\html-report\index.html
Specifications: 1 executed      0 passed        1 failed        0 skipped
Scenarios:      0 executed      0 passed        0 failed        0 skipped

Total time taken: 4.571s
Updates are available. Run `gauge update -c` for more info.
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! gauge-taiko-template@0.0.1 test.u: `gauge run specs/underwriter.spec`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the gauge-taiko-template@0.0.1 test.u script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.


### Versions
```
Paste `node` and `taiko` versions here.
```
![image](https://user-images.githubusercontent.com/25215979/87466117-bd283b80-c5e3-11ea-9eed-f08e37d08f88.png)

![image](https://user-images.githubusercontent.com/25215979/87466647-a33b2880-c5e4-11ea-8587-8dcb2d9118f1.png)

The debug log output is below:
0 info it worked if it ends with ok
1 verbose cli [ 'C:\\Program Files\\nodejs\\node.exe',
1 verbose cli   'C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js',
1 verbose cli   'run',
1 verbose cli   'test.u' ]
2 info using npm@6.4.1
3 info using node@v10.10.0
4 verbose run-script [ 'pretest.u', 'test.u', 'posttest.u' ]
5 info lifecycle gauge-taiko-template@0.0.1~pretest.u: gauge-taiko-template@0.0.1
6 info lifecycle gauge-taiko-template@0.0.1~test.u: gauge-taiko-template@0.0.1
7 verbose lifecycle gauge-taiko-template@0.0.1~test.u: unsafe-perm in lifecycle true
8 verbose lifecycle gauge-taiko-template@0.0.1~test.u: PATH: C:\Program Files\nodejs\node_modules\npm\node_modules\npm-lifecycle\node-gyp-bin;d:\ProgramData\BondDevSB\TrunkServices\Bidscout_Pipeline\CI_CD\e2e_tests\cs-bidscout-web-config\node_modules\.bin;C:\Program Files (x86)\Python37-32\Scripts\;C:\Program Files (x86)\Python37-32\;C:\Program Files (x86)\Common Files\Oracle\Java\javapath;C:\Program Files\Microsoft MPI\Bin\;C:\Program Files\CA\webagent\win32\bin;C:\Program Files\CA\webagent\win64\bin;C:\Program Files\CA\webagent\win64\install_config_info\lib;C:\ProgramData\Oracle\Java\javapath;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\Strawberry\Perl_5.24.1.1_x64;C:\Program Files\Microsoft SQL Server\120\Tools\Binn\;C:\Program Files\Microsoft SQL Server\130\Tools\Binn\;C:\Program Files (x86)\Microsoft Emulator Manager\1.0\;C:\Program Files\dotnet\;C:\Program Files\TortoiseSVN\bin;C:\Program Files\SlikSvn\bin;C:\Users\BondBldDeploy\AppData\Roaming\npm;C:\Program Files\nodejs\node_modules\npm;C:\Program Files (x86)\Microsoft SDKs\TypeScript\1.0\;D:\programdata\BondDevSB\Leg3\bondbuilds\BondBuildProcess\NantBinaries;C:\Program Files (x86)\Microsoft SDKs\Windows\v10.0A\bin\NETFX 4.6.1 Tools;C:\blat;C:\Infozip;C:\Program Files (x86)\AdminStudio\2015\Common\;C:\Program Files\Microsoft SQL Server\110\Tools\Binn\;C:\Program Files\Cloud Foundry;C:\Program Files (x86)\Microsoft VS Code\bin;C:\Program Files\Git\cmd;C:\Program Files\Git\bin;D:\ProgramData\BondDevSB\bondbuilds\BondBuildProcess\NantBinaries;C:\Program Files\Git\usr\bin;C:\Windows\SysWOW64\;C:\Windows\SysWOW64\Wbem;C:\Windows\SysWOW64\WindowsPowerShell\v1.0;C:\Program Files\Microsoft Windows Performance Toolkit;C:\Program Files\nodejs\;C:\Program Files\Microsoft SQL Server\Client SDK\ODBC\170\Tools\Binn\;C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\MSBuild\15.0\Bin;C:\Program Files\dotnet\sdk\2.2.105;C:\Program Files (x86)\Gauge;C:\ProgramData\chocolatey\bin;C:\tools\BCURRAN3;C:\Program Files\Puppet Labs\Puppet\bin;D:\GaugeSoftware\gauge;C:\Users\bondCNSQA_d\.dotnet\tools;C:\Users\bondCNSQA_d\AppData\Roaming\npm;D:\GaugeSoftware
9 verbose lifecycle gauge-taiko-template@0.0.1~test.u: CWD: d:\ProgramData\BondDevSB\TrunkServices\Bidscout_Pipeline\CI_CD\e2e_tests\cs-bidscout-web-config
10 silly lifecycle gauge-taiko-template@0.0.1~test.u: Args: [ '/d /s /c', 'gauge run specs/underwriter.spec' ]
11 silly lifecycle gauge-taiko-template@0.0.1~test.u: Returned: code: 1  signal: null
12 info lifecycle gauge-taiko-template@0.0.1~test.u: Failed to exec test.u script
13 verbose stack Error: gauge-taiko-template@0.0.1 test.u: `gauge run specs/underwriter.spec`
13 verbose stack Exit status 1
13 verbose stack     at EventEmitter.<anonymous> (C:\Program Files\nodejs\node_modules\npm\node_modules\npm-lifecycle\index.js:301:16)
13 verbose stack     at EventEmitter.emit (events.js:182:13)
13 verbose stack     at ChildProcess.<anonymous> (C:\Program Files\nodejs\node_modules\npm\node_modules\npm-lifecycle\lib\spawn.js:55:14)
13 verbose stack     at ChildProcess.emit (events.js:182:13)
13 verbose stack     at maybeClose (internal/child_process.js:962:16)
13 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:251:5)
14 verbose pkgid gauge-taiko-template@0.0.1
15 verbose cwd d:\ProgramData\BondDevSB\TrunkServices\Bidscout_Pipeline\CI_CD\e2e_tests\cs-bidscout-web-config
16 verbose Windows_NT 6.3.9600
17 verbose argv "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js" "run" "test.u"
18 verbose node v10.10.0
19 verbose npm  v6.4.1
20 error code ELIFECYCLE
21 error errno 1
22 error gauge-taiko-template@0.0.1 test.u: `gauge run specs/underwriter.spec`
22 error Exit status 1
23 error Failed at the gauge-taiko-template@0.0.1 test.u script.
23 error This is probably not a problem with npm. There is likely additional logging output above.
24 verbose exit [ 1, true ]

--- a/lib/handlers/fetchHandler.js
+++ b/lib/handlers/fetchHandler.js
@@ -48,6 +48,7 @@ const addExtraHeadersToRequest = (p) => {
       options.headers[headerName] = headers[headerName];
     }
   }
+  options.headers = headerArray(options.headers);
   return options;
 };
 
@@ -251,9 +252,13 @@ const resetInterceptors = () => {
   userEnabledIntercept = false;
 };
 
-const setHTTPHeaders = (headers, url) => {
+const setHTTPHeaders = async (headers, url) => {
   let host = extractHostName(url);
   headersMap.set(host, headers);
+  if (!userEnabledIntercept) {
+    userEnabledIntercept = true;
+    await enableFetchIntercept();
+  }
 };
 
 module.exports = {

--- a/lib/handlers/fetchHandler.js
+++ b/lib/handlers/fetchHandler.js
@@ -48,7 +48,9 @@ const addExtraHeadersToRequest = (p) => {
       options.headers[headerName] = headers[headerName];
     }
   }
-  options.headers = headerArray(options.headers);
+  if (options.headers) {
+    options.headers = headerArray(options.headers);
+  }
   return options;
 };
 

--- a/lib/taiko.js
+++ b/lib/taiko.js
@@ -908,7 +908,7 @@ module.exports.goto = async (
     url = 'http://' + url;
   }
   if (options.headers) {
-    fetchHandler.setHTTPHeaders(options.headers, url);
+    await fetchHandler.setHTTPHeaders(options.headers, url);
   }
   options = setNavigationOptions(options);
   let response;

--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "$schema": "http://json.schemastore.org/package",
   "name": "taiko",
-  "version": "1.0.14",
+  "version": "1.0.15",
   "description": "Taiko is a Node.js library for automating Chromium based browsers",
   "main": "bin/taiko.js",
   "bin": {

--- a/test/functional-tests/specs/pageActions.spec
+++ b/test/functional-tests/specs/pageActions.spec
@@ -99,3 +99,7 @@
 * Navigate to relative path "./specs/data/samePageNavigation.html#gauge-navigation"
 * Navigate to relative path "./specs/data/samePageNavigation.html#gauge-navigation"
 
+## Goto with basic authentication
+* Navigate to "http://localhost:3001/basic_auth" with basic auth "admin" and "admin"
+* Assert text "Congratulations! You must have the proper credentials." exists on the page.
+

--- a/test/functional-tests/tests/htmlElementAPI.ts
+++ b/test/functional-tests/tests/htmlElementAPI.ts
@@ -36,6 +36,14 @@ export default class HtmlElementAPI {
     await goto(url);
   }
 
+  @Step('Navigate to <url> with basic auth <username> and <password>')
+  public async navigateWithBasicAuth(url, username, password) {
+    const encodedCredentials = Buffer.from(`${username}:${password}`).toString('base64');
+    await goto(url, {
+      headers: { Authorization: `Basic ${encodedCredentials}` },
+    });
+  }
+
   @Step('Ensure Drop down <dropDownName> exists')
   public async dropdownExists(dropDownName) {
     const box = dropDown(dropDownName);

--- a/test/unit-tests/handlers/fetchHandler.test.js
+++ b/test/unit-tests/handlers/fetchHandler.test.js
@@ -10,6 +10,8 @@ describe('Fetch Handler', () => {
   before(() => {
     fetchHandler = rewire('../../../lib/handlers/fetchHandler');
     let fetch = {
+      enable: () => {},
+      requestPaused: () => {},
       continueRequest: async (options) => {
         continueInterceptedRequestOptions = options;
       },
@@ -27,12 +29,32 @@ describe('Fetch Handler', () => {
   });
 
   describe('http headers', () => {
-    let headersAndHost;
+    let headersAndHost, expectedHeaders;
     before(() => {
       headersAndHost = [
         [{ header1: 'header1 value' }, 'https://example.com'],
         [{ header3: 'header2 value' }, 'https://another-example.com'],
         [{ header4: 'header3 value' }, 'file://path/to/some/file'],
+      ];
+      expectedHeaders = [
+        [
+          {
+            name: 'header1',
+            value: 'header1 value',
+          },
+        ],
+        [
+          {
+            name: 'header3',
+            value: 'header2 value',
+          },
+        ],
+        [
+          {
+            name: 'header4',
+            value: 'header3 value',
+          },
+        ],
       ];
       headersAndHost.forEach((headerAndHost) => {
         fetchHandler.setHTTPHeaders(headerAndHost[0], headerAndHost[1]);
@@ -41,7 +63,7 @@ describe('Fetch Handler', () => {
     it('should set appropriate headers for a host', () => {
       headersAndHost.forEach((headerAndHost, index) => {
         const hostUrl = headerAndHost[1];
-        const headers = headerAndHost[0];
+        const headers = expectedHeaders[index];
         requestInterceptor({
           requestId: index,
           request: { url: hostUrl, headers: {} },

--- a/types/taiko/index.d.ts
+++ b/types/taiko/index.d.ts
@@ -35,7 +35,7 @@ declare module 'taiko' {
   }
 
   export interface NavigationOptions extends BasicNavigationOptions, EventOptions {
-    headers?: Map<string, string>;
+    headers?: object;
     waitForStart?: boolean;
   }
 

